icu4c/source/tools/toolutil/pkg_genc.cpp

```diff
@@ -847,7 +852,24 @@ getArchitecture(uint16_t *pCPU, uint16_t *pBits, UBool *pIsBigEndian, const char
 #   if defined(_M_IX86)
         *pCPU = IMAGE_FILE_MACHINE_I386;
 #   else
-        *pCPU = IMAGE_FILE_MACHINE_UNKNOWN;
+        // Linker for ClangCL doesn't handle IMAGE_FILE_MACHINE_UNKNOWN the same as
+        // linker for MSVC. Because of this optCpuArch is used to define the CPU
+        // architecture in that case. While _M_AMD64 and _M_ARM64 could be used,
+        // this would potentially be problematic when cross-compiling as this code
+        // would most likely be ran on host machine to generate the .obj file for
+        // the target architecture.
+#       if defined(__clang__)
+            if (strcmp(optCpuArch, "x64") == 0) {
+                *pCPU = IMAGE_FILE_MACHINE_AMD64;
+            } else if (strcmp(optCpuArch, "arm64") == 0) {
+                *pCPU = IMAGE_FILE_MACHINE_ARM64;
+            } else {
+                // This should never happen.
```

Comment:
In that case, I'd say that it'd be easier to debug if the process terminated with an error message if that which should never happen ever happened anyway.

Reply:
Makes sense. I've changed it.

Reply:
I added it based on your previous comment:

> In that case, I'd say that it'd be easier to debug if the process terminated with an error message if that which should never happen ever happened anyway.

Although I didn't think about it before you mentioned it, this log makes sense to me. For example, imagine someone makes a typo and passes `-c X64` or `-c amr64`, this log would help them fix it much quicker.

Reply:
> For example, imagine someone makes a typo and passes `-c X64` or `-c amr64`, this log would help them fix it much quicker.

Wait a second, your original comment here was _"This should never happen."_ but the user making a typo on the commandline is the total opposite of that, that's something that certainly is going to happen, sooner or later.

But verifying that the commandline arguments are correct should be done already in `main()` in `genccode.c`, in the same way that the arguments to `-a` are being verified there by calling the `checkAssemblyHeaderName()` helper function.

Reply:
I understand the confusion. My comment was misleading. Sorry for that. I checked the code now and saw the `checkAssemblyHeaderName` helper, I'll add that type of check to the `-c` option as well. Thanks for pointing that out. Once I add that the else branch should really _**never happen**_, and then the change you suggested with `std::terminate()` will be correct.